File: erlang/src/p017.erl
-module(p017).

-export([start/0]).

f() ->
  D = dict:from_list([
    {1, "one"},
    {2, "two"},
    {3, "three"},
    {4, "four"},
    {5, "five"},
    {6, "six"},
    {7, "seven"},
    {8, "eight"},
    {9, "nine"},
    {10, "ten"},
    {11, "eleven"},
    {12, "twelve"},
    {13, "thirteen"},
    {14, "fourteen"},
    {15, "fifteen"},
    {16, "sixteen"},
    {17, "seventeen"},
    {18, "eighteen"},
    {19, "nineteen"},
    {20, "twenty"},
    {30, "thirty"},
    {40, "forty"},
    {50, "fivty"},
    {60, "sixty"},
    {70, "seventy"},
    {80, "eighty"},
    {90, "ninety"},
    {100, "hundred"},
    {1000, "thousand"}
  ]),
  lists:sum([ length(strip(name(N, D))) || N <- lists:seq(1, 1000) ]).

strip(Word) ->
  re:replace(Word, "[\s-]", "", [global, {return, list}]).

name(N, D) when N =< 20 ->
  dict:fetch(N, D);
name(N, D) when N rem 10 == 0, N =< 90 ->
  dict:fetch(N, D);
name(N, D) when N rem 1000 == 0 ->
  dict:fetch(N div 1000, D) ++ " " ++ dict:fetch(1000, D);
name(N, D) when N rem 100 == 0 ->
  dict:fetch(N div 100, D) ++ " " ++ dict:fetch(100, D);
name(N, D) when N >= 100 ->
  dict:fetch(N div 100, D) ++ " " ++ dict:fetch(100, D) ++ " and " ++ name(N rem 100, D);
name(N, D) ->
  dict:fetch(N div 10 * 10, D) ++ "-" ++ dict:fetch(N rem 10, D).

start() ->
  io:format("~p ~n", [f()]).

File: go/p017.go
package main

import "fmt"

func main() {

    m := make(map[int]string)
    m[1] = "one"
    m[2] = "two"
    m[3] = "three"
    m[4] = "four"
    m[5] = "five"
    m[6] = "six"
    m[7] = "seven"
    m[8] = "eight"
    m[9] = "nine"
    m[10] = "ten"
    m[11] = "eleven"
    m[12] = "twelve"
    m[13] = "thirteen"
    m[14] = "fourteen"
    m[15] = "fifteen"
    m[16] = "sixteen"
    m[17] = "seventeen"
    m[18] = "eighteen"
    m[19] = "nineteen"
    m[20] = "twenty"
    m[30] = "thirty"
    m[40] = "forty"
    m[50] = "fivty"
    m[60] = "sixty"
    m[70] = "seventy"
    m[80] = "eighty"
    m[90] = "ninety"
    m[100] = "hundred"
    m[1000] = "thousand"

    count := 0
    for i := 1; i <= 1000; i++ {
       count += len(word(m, i))
    }
    fmt.Println(count)
}


func word(m map[int]string, n int) (string) {
    if n <= 20 {
        return m[n]
    } else if n % 10 == 0 && n <= 90 {
        return m[n]
    } else if n % 1000 == 0 {
        return fmt.Sprint(m[n / 1000], m[1000])
    } else if n % 100 == 0 {
        return fmt.Sprint(m[n / 100], m[100])
    } else if n >= 100 {
        return fmt.Sprint(m[n / 100], m[100], "and", word(m, n % 100))
    } else {
        return fmt.Sprint(word(m, n / 10 * 10), m[n % 10])
    }
}
